README.rst
********
Overview
********

.. start-badges

.. list-table::
    :stub-columns: 1

    * - docs
      - |docs|
    * - tests
      - | |gh_actions|
        | |codecov|
    * - package
      - | |version| |wheel|
        | |supported-versions|
        | |supported-implementations|

.. |docs| image:: https://readthedocs.org/projects/cbytesparse/badge/?style=flat
    :target: https://readthedocs.org/projects/cbytesparse
    :alt: Documentation Status

.. |gh_actions| image:: https://github.com/TexZK/cbytesparse/workflows/CI/badge.svg
    :alt: GitHub Actions Status
    :target: https://github.com/TexZK/cbytesparse

.. |codecov| image:: https://codecov.io/gh/TexZK/cbytesparse/branch/main/graphs/badge.svg?branch=main
    :alt: Coverage Status
    :target: https://codecov.io/github/TexZK/cbytesparse

.. |version| image:: https://img.shields.io/pypi/v/cbytesparse.svg
    :alt: PyPI Package latest release
    :target: https://pypi.org/project/cbytesparse/

.. |wheel| image:: https://img.shields.io/pypi/wheel/cbytesparse.svg
    :alt: PyPI Wheel
    :target: https://pypi.org/project/cbytesparse/

.. |supported-versions| image:: https://img.shields.io/pypi/pyversions/cbytesparse.svg
    :alt: Supported versions
    :target: https://pypi.org/project/cbytesparse/

.. |supported-implementations| image:: https://img.shields.io/pypi/implementation/cbytesparse.svg
    :alt: Supported implementations
    :target: https://pypi.org/project/cbytesparse/

.. end-badges


Library to handle sparse bytes within a virtual memory space.

* Free software: BSD 2-Clause License


Objectives
==========

This library aims to provide utilities to work with a `virtual memory`, which
consists of a virtual addressing space where sparse `chunks` of data can be
stored.

In order to be easy to use, its interface should be close to that of a
``bytearray``, which is the closest pythonic way to store dynamic data.
The main downside of a ``bytearray`` is that it requires a contiguous data
allocation starting from address 0. This is not good when sparse data have to
be stored, such as when emulating the addressing space of a generic
microcontroller.

The main idea is to provide a ``bytearray``-like class with the possibility to
internally hold the sparse `blocks` of data.
A `block` is ideally a tuple ``(start, data)`` where `start` is the start
address and `data` is the container of data items (e.g. ``bytearray``).
The length of the block is ``len(data)``.
Those blocks are usually not overlapping nor contiguous, and sorted by start
address.


Python implementation
=====================

This library is the Cython complement to the Python implementation provided by
the ``bytesparse`` Python package.
Please refer to its own documentation for more details.

The ``bytesparse`` package provides the following virtual memory types:

* ``bytesparse.Memory``, a generic virtual memory with infinite address range.
* ``bytesparse.bytesparse``, a subclass behaving more like ``bytearray``.

All the implementations inherit the behavior of
``collections.abc.MutableSequence`` and ``collections.abc.MutableMapping``.
Please refer to `the collections.abc reference manual
<https://docs.python.org/3/library/collections.abc.html>`_ for more information
about the interface API methods and capabilities.


Cython implementation
=====================

The library provides an experimental `Cython` implementation. It tries to
mimic the same algorithms of the Python implementation, while exploiting the
speedup of compiled `C` code.

Beware that the Cython implementation is meant to be potentially faster than
the pure Python one, but there might be even faster `ad-hoc` implementations
of virtual memory highly optimized for the underlying hardware.

The addressing space is limited to that of an ``uint_fast64_t``, so it is not
possible to have an infinite addressing space, nor negative addresses.
To keep the implementation code simple enough, the highest address (i.e.
``0xFFFFFFFFFFFFFFFF``) is reserved.

Block data chunks cannot be greater than the maximum ``ssize_t`` value
(typically half of the addressing space).

The Cython implementation is optional, and potentially useful only when the
Python implementation seems too slow for the user's algorithms, within the
limits stated above.

If in doubt about using the Cython implementation, just stick with the Python
one, which is much easier to integrate and debug.

More details can be found within ``cbytesparse.c``.


Examples
========

Here's a quick usage example of ``bytesparse`` objects:

>>> from cbytesparse import Memory
>>> from cbytesparse import bytesparse
>>> # ----------------------------------------------------------------
>>> m = bytesparse(b'Hello, World!')  # creates from bytes
>>> len(m)  # total length
13
>>> str(m)  # string representation, with bounds and data blocks
"<[[0, b'Hello, World!']]>"
>>> bytes(m)  # exports as bytes
b'Hello, World!'
>>> m.to_bytes()  # exports the whole range as bytes
b'Hello, World!'
>>> # ----------------------------------------------------------------
>>> m.extend(b'!!')  # more emphasis!!!
>>> bytes(m)
b'Hello, World!!!'
>>> # ----------------------------------------------------------------
>>> i = m.index(b',')  # gets the address of the comma
>>> m[:i] = b'Ciao'  # replaces 'Hello' with 'Ciao'
>>> bytes(m)
b'Ciao, World!!!'
>>> # ----------------------------------------------------------------
>>> i = m.index(b',')  # gets the address of the comma
>>> m.insert(i, b'ne')  # inserts 'ne' to make 'Ciaone' ("big ciao")
>>> bytes(m)
b'Ciaone, World!!!'
>>> # ----------------------------------------------------------------
>>> i = m.index(b',')  # gets the address of the comma
>>> m[(i - 2):i] = b' ciao'  # makes 'Ciaone' --> 'Ciao ciao'
>>> bytes(m)
b'Ciao ciao, World!!!'
>>> # ----------------------------------------------------------------
>>> m.pop()  # less emphasis --> 33 == ord('!')
33
>>> bytes(m)
b'Ciao ciao, World!!'
>>> # ----------------------------------------------------------------
>>> del m[m.index(b'l')]  # makes 'World' --> 'Word'
>>> bytes(m)
b'Ciao ciao, Word!!'
>>> # ----------------------------------------------------------------
>>> m.popitem()  # less emphasis --> pops 33 (== '!') at address 16
(16, 33)
>>> bytes(m)
b'Ciao ciao, Word!'
>>> # ----------------------------------------------------------------
>>> m.remove(b' ciao')  # self-explanatory
>>> bytes(m)
b'Ciao, Word!'
>>> # ----------------------------------------------------------------
>>> i = m.index(b',')  # gets the address of the comma
>>> m.clear(start=i, endex=(i + 2))  # makes empty space between the words
>>> m.to_blocks()  # exports as data block list
[[0, b'Ciao'], [6, b'Word!']]
>>> m.contiguous  # multiple data blocks (emptiness inbetween)
False
>>> m.content_parts  # two data blocks
2
>>> m.content_size  # excluding emptiness
9
>>> len(m)  # including emptiness
11
>>> # ----------------------------------------------------------------
>>> m.flood(pattern=b'.')  # replaces emptiness with dots
>>> bytes(m)
b'Ciao..Word!'
>>> m[-2]  # 100 == ord('d')
100
>>> # ----------------------------------------------------------------
>>> m.peek(-2)  # 100 == ord('d')
100
>>> m.poke(-2, b'k')  # makes 'Word' --> 'Work'
>>> bytes(m)
b'Ciao..Work!'
>>> # ----------------------------------------------------------------
>>> m.crop(start=m.index(b'W'))  # keeps 'Work!'
>>> m.to_blocks()
[[6, b'Work!']]
>>> m.span  # address range of the whole memory
(6, 11)
>>> m.start, m.endex  # same as above
(6, 11)
>>> # ----------------------------------------------------------------
>>> m.bound_span = (2, 10)  # sets memory address bounds
>>> str(m)
"<2, [[6, b'Work']], 10>"
>>> m.to_blocks()
[[6, b'Work']]
>>> # ----------------------------------------------------------------
>>> m.shift(-6)  # shifts to the left; NOTE: address bounds will cut 2 bytes!
>>> m.to_blocks()
[[2, b'rk']]
>>> str(m)
"<2, [[2, b'rk']], 10>"
>>> # ----------------------------------------------------------------
>>> a = bytesparse(b'Ma')
>>> a.write(0, m)  # writes [2, b'rk'] --> 'Mark'
>>> a.to_blocks()
[[0, b'Mark']]
>>> # ----------------------------------------------------------------
>>> b = Memory.from_bytes(b'ing', offset=4)
>>> b.to_blocks()
[[4, b'ing']]
>>> # ----------------------------------------------------------------
>>> a.write(0, b)  # writes [4, b'ing'] --> 'Marking'
>>> a.to_blocks()
[[0, b'Marking']]
>>> # ----------------------------------------------------------------
>>> a.reserve(4, 2)  # inserts 2 empty bytes after 'Mark'
>>> a.to_blocks()
[[0, b'Mark'], [6, b'ing']]
>>> # ----------------------------------------------------------------
>>> a.write(4, b'et')  # --> 'Marketing'
>>> a.to_blocks()
[[0, b'Marketing']]
>>> # ----------------------------------------------------------------
>>> a.fill(1, -1, b'*')  # fills asterisks between the first and last letters
>>> a.to_blocks()
[[0, b'M*******g']]
>>> # ----------------------------------------------------------------
>>> v = a.view(1, -1)  # creates a memory view spanning the asterisks
>>> v[::2] = b'1234'  # replaces even asterisks with numbers
>>> a.to_blocks()
[[0, b'M1*2*3*4g']]
>>> a.count(b'*')  # counts all the asterisks
3
>>> v.release()  # release memory view
>>> # ----------------------------------------------------------------
>>> c = a.copy()  # creates a (deep) copy
>>> c == a
True
>>> c is a
False
>>> # ----------------------------------------------------------------
>>> del a[a.index(b'*')::2]  # deletes every other byte from the first asterisk
>>> a.to_blocks()
[[0, b'M1234']]
>>> # ----------------------------------------------------------------
>>> a.shift(3)  # moves away from the trivial 0 index
>>> a.to_blocks()
[[3, b'M1234']]
>>> list(a.keys())
[3, 4, 5, 6, 7]
>>> list(a.values())
[77, 49, 50, 51, 52]
>>> list(a.items())
[(3, 77), (4, 49), (5, 50), (6, 51), (7, 52)]
>>> # ----------------------------------------------------------------
>>> c.to_blocks()  # reminder
[[0, b'M1*2*3*4g']]
>>> c[2::2] = None  # clears (empties) every other byte from the first asterisk
>>> c.to_blocks()
[[0, b'M1'], [3, b'2'], [5, b'3'], [7, b'4']]
>>> list(c.intervals())  # lists all the block ranges
[(0, 2), (3, 4), (5, 6), (7, 8)]
>>> list(c.gaps())  # lists all the empty ranges
[(None, 0), (2, 3), (4, 5), (6, 7), (8, None)]
>>> # ----------------------------------------------------------------
>>> c.flood(pattern=b'xy')  # fills empty spaces
>>> c.to_blocks()
[[0, b'M1x2x3x4']]
>>> # ----------------------------------------------------------------
>>> t = c.cut(c.index(b'1'), c.index(b'3'))  # cuts an inner slice
>>> t.to_blocks()
[[1, b'1x2x']]

CHANGELOG.rst
Changelog
=========

0.0.6 (2023-02-18)
------------------

* Following the ``bytesparse`` Python package, version ``0.0.6``.
* Added automatic wheel builder within GitHub Actions CI.
* Added *byte-like* methods and inplace editing.
* Added support to Python 3.11, removed 3.6.
* Added some minor features.
* Improved documentation.
* Improved testing.
* Improved repository layout (``pyproject.toml``).
* Minor fixes.


0.0.5 (2022-02-22)
------------------

* Following the ``bytesparse`` Python package, version ``0.0.5``.
* Added ``bytesparse`` class, closer to ``bytearray`` than ``Memory``.
* Added registration to the abstract base classes.
* Added missing abstract and ported methods.
* Added cut feature.
* Added more helper methods.
* Fixed values iteration.
* Improved extraction performance.
* Improved testing.


0.0.4 (2022-01-09)
------------------

* Following the ``bytesparse`` Python package, version ``0.0.4``.
* Refactored current implementation as the ``c`` sub-module.
* Removed experimental backup feature.
* Added dedicated methods to backup/restore mutated state.
* Fixed some write/insert bugs.
* Fixed some trim/bound bugs.
* Methods sorted by name.
* Removed useless functions.


0.0.2 (2022-01-03)
------------------

* Forced extension compilation.
* Using explicit factory methods instead of constructor arguments.
* Added block collapsing helper function.
* Minor fixes.
* Improved test suite.


0.0.1 (2021-12-27)
------------------

* First release on PyPI.


src/cbytesparse/base.py
# Copyright (c) 2020-2023, Andrea Zoppi.
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
# 1. Redistributions of source code must retain the above copyright notice,
#    this list of conditions and the following disclaimer.
#
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

r"""Common stuff, shared across modules."""

import abc
import collections.abc
from typing import Any
from typing import ByteString
from typing import Callable
from typing import Iterable
from typing import List
from typing import Optional
from typing import Tuple
from typing import Union

try:
    from typing import TypeAlias
except ImportError:  # pragma: no cover
    TypeAlias = Any  # Python < 3.10

BytesLike: TypeAlias = Union[ByteString, memoryview]
BytesFactory: TypeAlias = Callable[..., BytesLike]


def _inherit_extend_docstring(src_class):  # pragma: no cover
    def wrapper(dst_method):  # pragma: no cover
        if hasattr(src_class, dst_method.__name__):
            docstring_after = r"""

Warnings:
    This method documentation is just a stub, copied directly from
    :meth:`{0}.{1}`. It may be incomplete, or with different arguments.
    The behavior should be very similar though.
            """.format(src_class.__name__, dst_method.__name__)

            src_method = getattr(src_class, dst_method.__name__)
            dst_method.__doc__ = src_method.__doc__ + docstring_after
        return dst_method
    return wrapper


class BaseBytesMethods(ByteString, collections.abc.Sequence):
    r"""Provides useful methods to a byte buffer.

    Python's :obj:`memoryview` and most *byte-like* objects do not provide many
    useful methods found instead within the :obj:`bytes` or :obj:`str` APIs.

    This wrapper class adds a low-level implementation of those methods to
    anything supporting the *buffer protocol*.

    Arguments:
        wrapped (*byte-like*): The target object supporting the
            *buffer protocol*.

    Examples:
        >>> from cbytesparse import BytesMethods
        >>> import numpy
        >>> numbers = list(b'ABC')
        >>> numbers
        [65, 66, 67]
        >>> data = numpy.array(numbers, dtype=numpy.ubyte)
        >>> data
        array([65, 66, 67], dtype=uint8)
        >>> data.lower()  # noqa
        Traceback (most recent call last):
            ...
        AttributeError: 'numpy.ndarray' object has no attribute 'lower'
        >>> wrapped = BytesMethods(data)  # noqa
        >>> bytes(wrapped.lower())
        b'abc'
        >>> wrapped = BytesMethods(memoryview(data))
        >>> bytes(wrapped.lower())
        b'abc'
    """

    @abc.abstractmethod
    def __bool__(
        self,
    ) -> bool:
        r"""Has any items.

        Returns:
            bool: Has any items.

        Examples:
            >>> from cbytesparse import BytesMethods

            >>> instance = BytesMethods(b'')
            >>> bool(instance)
            False

            >>> instance = BytesMethods(b'Hello, World!')
            >>> bool(instance)
            True
        """
        ...

    @abc.abstractmethod
    def __bytes__(
        self,
    ) -> bytes:
        r"""Creates a bytes clone.

        Returns:
            :obj:`bytes`: Cloned data.

        Examples:
            >>> from cbytesparse import BytesMethods

            >>> instance = BytesMethods(bytearray(b''))
            >>> bytes(instance)
            b''

            >>> instance = BytesMethods(bytearray(b'Hello, World!'))
            >>> bytes(instance)
            b'Hello, World!'
        """
        ...

    @abc.abstractmethod
    def __contains__(
        self,
        token: Union[BytesLike, int],
    ) -> bool:
        r"""Checks if some items are contained.

        Arguments:
            token (byte-like):
                Token to find.

        Returns:
            bool: Token is contained.

        Examples:
            >>> from cbytesparse import BytesMethods

            >>> instance = BytesMethods(b'Hello, World!')
            >>> b'World' in instance
            True
            >>> b'$' in instance
            False
            >>> ord('o') in instance
            True
            >>> ord('$') in instance
            False
        """
        ...

    @abc.abstractmethod
    def __delitem__(
        self,
        key: Any,
    ) -> None:
        ...

    @abc.abstractmethod
    def __eq__(
        self,
        other: Any,
    ) -> bool:
        r"""Equality comparison.

        Arguments:
            other (byte-like):
                Data to compare with `self`.

        Returns:
            bool: `self` is equal to `other`.

        Examples:
            >>> from cbytesparse import BytesMethods

            >>> data = bytearray(b'Hello, World!')
            >>> instance = BytesMethods(data)
            >>> instance == data
            True
            >>> instance == memoryview(data)
            True
            >>> instance == b'Something else'
            False
        """
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(bytes)  # TODO
    def __ge__(
        self,
        other: Any,
    ) -> bool:
        ...

    @abc.abstractmethod
    def __getitem__(
        self,
        key: Any,
    ) -> Any:
        r"""Gets data.

        Arguments:
            key (slice or int):
                Selection range or address.
                If it is a :obj:`slice` with bytes-like `step`, the latter is
                interpreted as the filling pattern.

        Returns:
            items: Items from the requested range.

        Examples:
            >>> from cbytesparse import BytesMethods

            >>> instance = BytesMethods(b'Hello, World!')
            >>> instance[7]  # -> ord('W') = 87
            121
            >>> bytes(instance[:3])
            b'Hel'
            >>> bytes(instance[3:10])
            b'lo, Wor'
            >>> bytes(instance[-1:])
            b'!'
            >>> bytes(instance[2:10:3])
            b'l,o'
            >>> bytes(instance[3:10:2])
            b'l,Wr'
        """
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(bytes)  # TODO
    def __gt__(
        self,
        other: Any,
    ) -> bool:
        r"""
        See Also:
            :meth:`bytes.__gt__`
        """
        ...

    @abc.abstractmethod
    def __init__(
        self,
        wrapped: Optional[BytesLike],
    ):
        ...

    @abc.abstractmethod
    def __iter__(
        self,
    ) -> Iterable[int]:
        r"""Iterates over values.

        Yields:
            int: Value as byte integer.

        Examples:
            >>> from cbytesparse import BytesMethods

            >>> instance = BytesMethods(b'Hello, World!')
            >>> list(instance)
            [72, 101, 108, 108, 111, 44, 32, 87, 111, 114, 108, 100, 33]
        """
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(bytes)  # TODO
    def __le__(
        self,
        other: Any,
    ) -> bool:
        ...

    @abc.abstractmethod
    def __len__(
        self,
    ) -> int:
        r"""Actual length.

        Computes the actual length of the wrapped data object.

        Returns:
            int: Data length.

        Examples:
            >>> from cbytesparse import BytesMethods

            >>> instance = BytesMethods(b'')
            >>> len(instance)
            0

            >>> instance = BytesMethods(bytes(7))
            >>> len(instance)
            7

            >>> instance = BytesMethods(memoryview(b'Hello, World!'))
            >>> len(instance)
            13
        """
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(bytes)  # TODO
    def __lt__(
        self,
        other: Any,
    ) -> bool:
        ...

    @abc.abstractmethod
    def __ne__(
        self,
        other: Any,
    ) -> bool:
        r"""Inquality comparison.

        Arguments:
            other (byte-like):
                Data to compare with `self`.

        Returns:
            bool: `self` is not equal to `other`.

        Examples:
            >>> from cbytesparse import BytesMethods

            >>> data = bytearray(b'Hello, World!')
            >>> instance = BytesMethods(data)
            >>> instance != data
            False
            >>> instance != memoryview(data)
            False
            >>> instance != b'Something else'
            True
        """
        ...

    @abc.abstractmethod
    def __reversed__(
        self,
    ) -> Iterable[int]:
        r"""Iterates over values, reversed order.

        Yields:
            int: Value as byte integer.

        Examples:
            >>> from cbytesparse import BytesMethods

            >>> instance = BytesMethods(b'Hello, World!')
            >>> list(instance)
            [72, 101, 108, 108, 111, 44, 32, 87, 111, 114, 108, 100, 33]
            >>> list(reversed(instance))
            [33, 100, 108, 114, 111, 87, 32, 44, 111, 108, 108, 101, 72]
        """
        ...

    @abc.abstractmethod
    def __setitem__(
        self,
        key: Any,
        value: Any,
    ) -> None:
        ...

    @abc.abstractmethod
    def __sizeof__(
        self,
    ) -> int:
        r"""int: Allocated byte size."""
        ...

    @property
    @abc.abstractmethod
    def c_contiguous(
        self,
    ) -> bool:
        r"""bool: Contiguous array, *C* language style."""
        ...

    @abc.abstractmethod
    def capitalize(
        self,
    ) -> 'BaseBytesMethods':
        r"""byte-like: First character capitalized, the rest lowercase."""
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(bytes)  # TODO
    def center(
        self,
        width: int,
        fillchar: BytesLike = b' ',
        factory: BytesFactory = bytes,
    ) -> BytesLike:
        ...

    @abc.abstractmethod
    def contains(
        self,
        token: Union[BytesLike, int],
        start: Optional[int] = None,
        endex: Optional[int] = None,
    ) -> bool:
        r"""Contains a substring.

        Arguments:
            token (byte-like):
                Token to search.

            start (int):
                Inclusive start of the searched range, ``None`` to ignore.

            endex (int):
                Exclusive end of the searched range, ``None`` to ignore.

        Returns:
            bool: Token contained within the wrapped object.

        See Also:
            :meth:`__contains__`

        Examples:
            >>> from cbytesparse import BytesMethods

            >>> instance = BytesMethods(b'Hello, World!')
            >>> instance.contains(b'World')
            True
            >>> instance.contains(b'$')
            False
            >>> instance.contains(ord('o'))
            True
            >>> instance.contains(ord('$'))
            False
            >>> instance.contains(b'Hello', endex=10)
            True
            >>> instance.contains(b'Hello', endex=3)
            False
            >>> instance.contains(b'World', start=3)
            True
            >>> instance.contains(b'World', start=10)
            False
            >>> instance.contains(b',', start=3, endex=10)
            True
            >>> instance.contains(b',', start=8, endex=10)
            False
        """
        ...

    @property
    @abc.abstractmethod
    def contiguous(
        self,
    ) -> bool:
        r"""bool: Contiguous array."""
        ...

    @abc.abstractmethod
    def count(
        self,
        token: Union[BytesLike, int],
        start: Optional[int] = None,
        endex: Optional[int] = None,
    ) -> int:
        r"""Counts token occurrences.

        Arguments:
            token (byte-like):
                Token to count.

            start (int):
                Inclusive start of the searched range, ``None`` to ignore.

            endex (int):
                Exclusive end of the searched range, ``None`` to ignore.

        Returns:
            int: The number of items equal to `token`.

        Examples:
            >>> from cbytesparse import BytesMethods

            >>> instance = BytesMethods(b'Hello, World!')
            >>> instance.count(b'o')
            2
            >>> instance.count(b'l')
            3
            >>> instance.count(b'll')
            3
            >>> instance.count(b'World')
            1
            >>> instance.count(b'$')
            0
            >>> instance.count(ord('o'))
            1
            >>> instance.count(ord('$'))
            0
            >>> instance.count(b'Hello', endex=10)
            1
            >>> instance.count(b'Hello', endex=3)
            0
            >>> instance.count(b'World', start=3)
            1
            >>> instance.count(b'World', start=10)
            0
            >>> instance.count(b',', start=3, endex=10)
            1
            >>> instance.count(b',', start=8, endex=10)
            0
        """
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(bytes)  # TODO
    def decode(
        self,
        encoding: str = 'utf-8',
        errors: str = 'strict',
    ) -> str:
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(bytes)  # TODO
    def endswith(
        self,
        token: BytesLike,
    ) -> bool:
        ...

    @property
    @abc.abstractmethod
    def f_contiguous(
        self,
    ) -> bool:
        r"""bool: Contiguous array, *Fortran* language style."""
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(bytes)  # TODO
    def find(
        self,
        token: Union[BytesLike, int],
        start: Optional[int] = None,
        endex: Optional[int] = None,
    ) -> int:
        ...

    @property
    @abc.abstractmethod
    @_inherit_extend_docstring(memoryview)  # TODO
    def format(
        self,
    ) -> str:
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(bytes)  # TODO
    def index(
        self,
        token: Union[BytesLike, int],
        start: Optional[int] = None,
        endex: Optional[int] = None,
    ) -> int:
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(bytes)  # TODO
    def isalnum(
        self,
    ) -> bool:
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(bytes)  # TODO
    def isalpha(
        self,
    ) -> bool:
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(bytes)  # TODO
    def isascii(
        self,
    ) -> bool:
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(str)  # TODO
    def isdecimal(
        self,
    ) -> bool:
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(bytes)  # TODO
    def isdigit(
        self,
    ) -> bool:
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(str)  # TODO
    def isidentifier(
        self,
    ) -> bool:
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(bytes)  # TODO
    def islower(
        self,
    ) -> bool:
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(str)  # TODO
    def isnumeric(
        self,
    ) -> bool:
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(str)  # TODO
    def isprintable(
        self,
    ) -> bool:
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(bytes)  # TODO
    def isspace(
        self,
    ) -> bool:
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(bytes)  # TODO
    def istitle(
        self,
    ) -> bool:
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(bytes)  # TODO
    def isupper(
        self,
    ) -> bool:
        ...

    @property
    @abc.abstractmethod
    @_inherit_extend_docstring(memoryview)  # TODO
    def itemsize(
        self,
    ) -> int:
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(bytes)  # TODO
    def ljust(
        self,
        width: int,
        fillchar: BytesLike = b' ',
        factory: BytesFactory = bytes,
    ) -> BytesLike:
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(bytes)  # TODO
    def lstrip(
        self,
        chars: Optional[BytesLike] = None,
        factory: BytesFactory = bytes,
    ) -> BytesLike:
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(bytes)  # TODO
    def lower(
        self,
    ) -> 'BaseBytesMethods':
        ...

    @staticmethod
    @abc.abstractmethod
    @_inherit_extend_docstring(bytes)  # TODO
    def maketrans(
        chars_from: BytesLike,
        chars_to: BytesLike,
    ) -> bytes:
        ...

    @property
    @abc.abstractmethod
    @_inherit_extend_docstring(memoryview)  # TODO
    def nbytes(
        self,
    ) -> int:
        ...

    @property
    @abc.abstractmethod
    @_inherit_extend_docstring(memoryview)  # TODO
    def ndim(
        self,
    ) -> int:
        ...

    @property
    @abc.abstractmethod
    @_inherit_extend_docstring(memoryview)  # TODO
    def obj(
        self,
    ) -> Optional[BytesLike]:
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(bytes)  # TODO
    def partition(
        self,
        sep: BytesLike,
        factory: BytesFactory = bytes,
    ) -> Tuple[BytesLike, BytesLike, BytesLike]:
        ...

    @property
    @abc.abstractmethod
    @_inherit_extend_docstring(memoryview)  # TODO
    def readonly(
        self,
    ) -> bool:
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(memoryview)  # TODO
    def release(
        self,
    ) -> None:
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(bytes)  # TODO
    def removeprefix(
        self,
        prefix: BytesLike,
        factory: BytesFactory = bytes,
    ) -> BytesLike:
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(bytes)  # TODO
    def removesuffix(
        self,
        suffix: BytesLike,
        factory: BytesFactory = bytes,
    ) -> BytesLike:
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(bytes)  # TODO
    def replace(
        self,
        old: BytesLike,
        new: BytesLike,
        count: Optional[int] = None,
        start: Optional[int] = None,
        endex: Optional[int] = None,
    ) -> 'BaseBytesMethods':
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(bytes)  # TODO
    def rfind(
        self,
        token: Union[BytesLike, int],
        start: Optional[int] = None,
        endex: Optional[int] = None,
    ) -> int:
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(bytes)  # TODO
    def rindex(
        self,
        token: Union[BytesLike, int],
        start: Optional[int] = None,
        endex: Optional[int] = None,
    ) -> int:
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(bytes)  # TODO
    def rjust(
        self,
        width: int,
        fillchar: BytesLike = b' ',
        factory: BytesFactory = bytes,
    ) -> BytesLike:
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(bytes)  # TODO
    def rpartition(
        self,
        sep: BytesLike,
        factory: BytesFactory = bytes,
    ) -> Tuple[BytesLike, BytesLike, BytesLike]:
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(bytes)  # TODO
    def rstrip(
        self,
        chars: Optional[BytesLike] = None,
        factory: BytesFactory = bytes,
    ) -> BytesLike:
        ...

    @property
    @abc.abstractmethod
    @_inherit_extend_docstring(memoryview)  # TODO
    def shape(
        self,
    ) -> Tuple[int]:
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(bytes)  # TODO
    def startswith(
        self,
        token: BytesLike,
    ) -> bool:
        r"""bool: Data beginning with *token*."""
        ...

    @property
    @abc.abstractmethod
    @_inherit_extend_docstring(memoryview)  # TODO
    def strides(
        self,
    ) -> Tuple[int]:
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(bytes)  # TODO
    def strip(
        self,
        chars: Optional[BytesLike] = None,
        factory: BytesFactory = bytes,
    ) -> BytesLike:
        ...

    @property
    @abc.abstractmethod
    @_inherit_extend_docstring(memoryview)  # TODO
    def suboffsets(
        self,
    ) -> Tuple:
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(bytes)  # TODO
    def swapcase(
        self,
    ) -> 'BaseBytesMethods':
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(bytes)  # TODO
    def title(
        self,
    ) -> 'BaseBytesMethods':
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(memoryview)  # TODO
    def tobytes(
        self,
    ) -> bytes:
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(memoryview)  # TODO
    def tolist(
        self,
    ) -> List[int]:
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(bytes)  # TODO
    def translate(
        self,
        table: BytesLike,
    ) -> 'BaseBytesMethods':
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(bytes)  # TODO
    def upper(
        self,
    ) -> 'BaseBytesMethods':
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(bytes)  # TODO
    def zfill(
        self,
        width: int,
        factory: BytesFactory = bytes,
    ) -> BytesLike:
        ...


class BaseInplaceView(BaseBytesMethods):
    r"""Provides inplace methods to a byte buffer.

    Standard Python *byte-like* objects (e.g. :obj:`memoryview` and
    :obj:`bytearray`) only provide a very small set of methods for
    *inplace* editing of their underlying bytes buffer.

    This wrapper class adds a low-level implementation of those methods to
    anything supporting the *buffer protocol* with a mutable buffer.

    Note:
        Editing support is only limited to the existing buffer items, i.e. the
        wrapper and its underlying buffer cannot be resized via the methods
        provided by this class, just like with standard memory views.

    Arguments:
        wrapped (*byte-like*): The target object supporting the
            *buffer protocol*.

    Examples:
        >>> from cbytesparse import InplaceView
        >>> buffer = bytearray(b'Hello, World!')
        >>> wrapped = InplaceView(buffer)
        >>> wrapped.lower()
        >>> buffer
        bytearray(b'hello, world!')
        >>> wrapped.replace(b'l', b'p')
        >>> buffer
        bytearray(b'heppo, worpd!')
        >>> wrapped.find(b'w')
        7
        >>> wrapped.rfind(b'o')
        8
        >>> wrapped.count(b'o')
        2
    """

    @abc.abstractmethod
    def __init__(
        self,
        wrapped: Optional[BytesLike],
    ):
        ...

    @abc.abstractmethod
    @_inherit_extend_docstring(memoryview)  # TODO
    def toreadonly(
        self,
    ) -> 'BaseInplaceView':
        ...
